pkg/sql/alter_index_visible.go

```diff
@@ -28,9 +33,49 @@ type alterIndexVisibleNode struct {
 func (p *planner) AlterIndexVisible(
 	ctx context.Context, n *tree.AlterIndexVisible,
 ) (planNode, error) {
-	return nil, unimplemented.Newf(
-		"Not Visible Index",
-		"altering an index to visible or not visible is not supported yet")
+	if err := checkSchemaChangeEnabled(
+		ctx,
+		p.ExecCfg(),
+		"ALTER INDEX VISIBILITY",
+	); err != nil {
+		return nil, err
+	}
+
+	// Check if the table actually exists. expandMutableIndexName returns the
+	// underlying table. If the index name is specified with a table name, it
+	// checks if the table exists. Otherwise, it searches the table using the
+	// index name. If no table is found, 1. IfExists is specified, tableDesc
+	// returned is nil. 2. No IfExists, error is returned.
+	_, tableDesc, err := expandMutableIndexName(ctx, p, &n.Index, !n.IfExists /* requireTable */)
+	if err != nil {
+		// Error if no table is found and IfExists is false.
+		return nil, err
+	}
+
+	if tableDesc == nil {
+		// No error if no table but IfExists is true.
+		return newZeroNode(nil /* columns */), nil
+	}
+
+	// Check if the index actually exists. FindIndexWithName returns the first
+	// catalog.Index in tableDesc.AllIndexes(). It returns error if none was found
+	// in the tableDesc. If the index name is specified with a table name, this is
+	// where we check if the index actually exists in the table.
+	idx, err := tableDesc.FindIndexWithName(string(n.Index.Index))
+	if err != nil {
+		if n.IfExists {
+			// Nothing needed if no index exists and IfExists is true.
+			return newZeroNode(nil /* columns */), nil
+		}
+		// Error if no index exists and IfExists is not specified.
+		return nil, pgerror.WithCandidateCode(err, pgcode.UndefinedObject)
```

Comment:
I suggest wrapping this in an assertion error here, `expandMutableIndexName` promises that if `tableDesc` is not nil, then the index exists, so an error here should never happen. Mentioning _that_ in the comments would be useful.